### PR TITLE
Dev/pointer offset

### PR DESCRIPTION
**Pull request by StephenHodgson**
# XRTK - Mixed Reality Toolkit Change Request

## Overview

Overhauled the `FocusProvider` to give more accurate data from the focused object.
Added a `SyncedTarget` to enable the cursor to be locked to an object while an object is in motion. (this is most noticeable in the `ManipulationHandler` in the associated SDK changes.

You can now grab something with a pointer and sync the pointer with the object's transform by the grab point while getting updated hit information that's not the object you picked up. This is useful when you're trying to fit the object into a space and it needs to collide with a spatial mesh or other GameObjects.

This PR also has a lot of other noise from cleaning up Extension Methods and formatting. 
I'll do my best to break these out into their own PRs when I can.

## Target of the change:

Is this enhancement for:

- Core

## Breaking Changes:

- `FocusDetails.Object ` -> `FocusDetails.HitObject`
- `IMixedRealityPointer.SyncPointerTargetPosition` -> `IMixedRealityPointer.SyncedTarget`
- Added lots of new properties to `IPointerResult`
- Removed the `IMixedRealityCursor.DefaultCursorDistance` as it was a duplicate. Use `IMixedRealityPointer.PointerExtent` instead.

### CODE REVIEW

File: XRTK-Core/Packages/com.xrtk.core/Interfaces/InputSystem/IPointerResult.cs [hunk 1]
```diff
@@ -78,5 +93,13 @@ public interface IPointerResult
...
         /// </summary>
         RaycastResult LastGraphicsRaycastResult { get; }
+
+        Vector3 GrabPointLocalSpace { get; }
+
+        Vector3 GrabPoint { get; }
+
+        Vector3 GrabPointOffsetLocalSpace { get; }
+
+        Vector3 GrabPointOffset { get; }
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Interfaces/InputSystem/IPointerResult.cs]: These properties need to be documented better

File: XRTK-Core/Packages/com.xrtk.core/Services/InputSystem/FocusProvider.cs [hunk 2]
```diff
@@ -251,10 +272,83 @@ public void UpdateHit(PointerHitResult hitResult)
...
+
+                    if (syncedPointerTarget != null)
+                    {
+                        Offset = CurrentPointerTarget.transform.position - GrabPointOffset;
+                        OffsetLocalSpace = CurrentPointerTarget.transform.InverseTransformPoint(Offset);
+
+                        //DebugUtilities.DrawPoint(EndPoint, Color.yellow);
+                        //Debug.DrawLine(EndPoint, GrabPoint, Color.yellow);
+
+                        //DebugUtilities.DrawPoint(Offset, Color.magenta);
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Services/InputSystem/FocusProvider.cs]: These need to be toggleable. Maybe using the raycaster debug flag?

File: XRTK-Core/Packages/com.xrtk.core/Utilities/Solvers/Solver.cs [hunk 3]
```diff
@@ -259,9 +231,9 @@ protected void UpdateTransformToGoal()
                 var rot = cachedTransform.rotation;
                 var scale = cachedTransform.localScale;
 
-                pos = SmoothTo(pos, GoalPosition, SolverHandler.DeltaTime, moveLerpTime);
-                rot = SmoothTo(rot, GoalRotation, SolverHandler.DeltaTime, rotateLerpTime);
-                scale = SmoothTo(scale, GoalScale, SolverHandler.DeltaTime, scaleLerpTime);
+                pos.SmoothTo(GoalPosition, SolverHandler.DeltaTime, moveLerpTime);
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Utilities/Solvers/Solver.cs]: I think all these places where I used the extension method actually broke these as their new values aren't being assigned back.

File: XRTK-Core/Packages/com.xrtk.core/Interfaces/InputSystem/IPointerResult.cs [hunk 4]
```diff
@@ -78,5 +79,9 @@ public interface IPointerResult
         /// The last raycast hit info for graphic raycast.
         /// </summary>
         RaycastResult LastGraphicsRaycastResult { get; }
+
+        Vector3 GrabPointLocalSpace { get; }
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Interfaces/InputSystem/IPointerResult.cs]: Needs documentation

File: XRTK-Core/Packages/com.xrtk.core/Interfaces/InputSystem/IPointerResult.cs [hunk 5]
```diff
@@ -78,5 +79,9 @@ public interface IPointerResult
...
+
+        Vector3 GrabPoint { get; }
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Interfaces/InputSystem/IPointerResult.cs]: Needs documentation

File: XRTK-Core/Packages/com.xrtk.core/Services/InputSystem/FocusProvider.cs [hunk 6]
```diff
@@ -251,10 +260,86 @@ public void UpdateHit(PointerHitResult hitResult)
...
+                    {
+                        GetCurrentTarget();
+                    }
+
+                    if (syncedPointerTarget != null)
+                    {
+                        GrabPoint = CurrentPointerTarget.transform.TransformPoint(GrabPointLocalSpace);
+                        GrabPointLocalSpace = CurrentPointerTarget.transform.InverseTransformPoint(GrabPoint);
+
+                        //DebugUtilities.DrawPoint(GrabPoint, Color.red);
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Services/InputSystem/FocusProvider.cs]: need a way to toggle this on and off for debugging

**Comment by StephenHodgson** [comment]
Ready for review

**Comment by SimonDarksideJ** [comment]
Build generating Mouse Pointer interface errors?

**Comment by StephenHodgson** [comment]
I'll check it out. Might have had a bad submodule merge

**Pull request merged**